codex/slots/builder.nim

```diff
@@ -252,24 +313,53 @@ proc new*(
     return failure("Block size must be divisable by cell size.")
 
   let
-    strategy = if strategy == nil:
+    strategy = if strategy.isNone:
       SteppedIndexingStrategy.new(
         0, manifest.blocksCount - 1, manifest.numSlots)
       else:
-        strategy
+        strategy.get
 
     # all trees have to be padded to power of two
     numBlockCells = manifest.blockSize.int div cellSize                       # number of cells per block
     blockPadBytes = newSeq[byte](numBlockCells.nextPowerOfTwoPad * cellSize)  # power of two padding for blocks
     numSlotLeafs = (manifest.blocksCount div manifest.numSlots)
     slotsPadLeafs = newSeqWith(numSlotLeafs.nextPowerOfTwoPad, Poseidon2Zero) # power of two padding for block roots
     rootsPadLeafs = newSeqWith(manifest.numSlots.nextPowerOfTwoPad, Poseidon2Zero)
+    blockEmptyDigest = ? Poseidon2Tree.digest(DefaultEmptyBlock & blockPadBytes, CellSize)
 
-  success SlotBuilder(
+  var self = SlotsBuilder(
     store: store,
     manifest: manifest,
     strategy: strategy,
     cellSize: cellSize,
     blockPadBytes: blockPadBytes,
     slotsPadLeafs: slotsPadLeafs,
-    rootsPadLeafs: rootsPadLeafs)
+    rootsPadLeafs: rootsPadLeafs,
+    blockEmptyDigest: blockEmptyDigest)
+
+  if manifest.verifiable:
+    if manifest.slotRoots.len == 0 or manifest.slotRoots.len != manifest.numSlots:
+      return failure "Manifest is verifiable but slot roots are missing or invalid."
+
+    let
+      slotRoot = ? Poseidon2Hash.fromBytes(
+        ( ? manifest.slotsRoot.mhash.mapFailure ).digestBytes.toArray32
+      ).toFailure
+
+      slotRoots = manifest.slotRoots.mapIt(
+        ? Poseidon2Hash.fromBytes(
+          ( ? it.mhash.mapFailure ).digestBytes.toArray32
+        ).toFailure
+      )
+
+    without root =? self.buildRootsTree(slotRoots).?root(), err:
+      error "Failed to build slot roots tree", err = err.msg
+      return failure(err)
+
+    if not bool(slotRoot == root): # TODO: `!=` doesn't work for SecretBool
```

Comment:
We have a fix for this in poseidon2. See https://github.com/codex-storage/nim-poseidon2/commit/eef2603c11eeb5069eaed6c0296f3a1c5d23aad1
Used: https://github.com/codex-storage/nim-codex/blob/1836c2476c4ac2822f8274f4b2530e07e53c9cc4/codex/proof/datasamplerstarter.nim#L80